src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,98 @@
+import org.springframework.stereotype.Service;
+
+@RequiredArgsConstructor
+@Service
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final CustomUserDetailsService userDetailsService;
+    private final CartRepository cartRepository;
+    private final CartMapper cartMapper;
+    private final CartItemMapper cartItemMapper;
+    private final CartItemRepository cartItemRepository;
+    private final BookRepository bookRepository;
+
+    @Override
+    public ShoppingCartDto getCart(Authentication authentication) {
+        Long userId = ((User) userDetailsService
+                .loadUserByUsername(authentication.getName())).getId();
+        ShoppingCart cart = cartRepository.findById(userId)
+                .orElseThrow(() -> new NoSuchElementException("No cart found"));
+        return cartMapper.toCartDto(cart);
+    }
+
+    @Override
+    public ShoppingCartDto createCartItem(
+            CreateCartItemDto createCartItemDto,
+            Authentication authentication
+    ) {
+        Long bookId = createCartItemDto.getBookId();
+        Book bookById = bookRepository.findById(bookId)
+                .orElseThrow(() -> new NoSuchElementException("Book not found"));
+        Long userId = ((User) userDetailsService
+                .loadUserByUsername(authentication.getName())).getId();
+        CartItem existingCartItem = cartItemRepository.findByCartIdAndBookId(userId, bookId);
+        if (existingCartItem != null) {
+            throw new RuntimeException("You can not add more than one cart item " +
+                    "type to the your cart. You can update its quantity");
+        }
+        CartItem cartItem = cartItemMapper.toEntity(createCartItemDto);
+        cartItem.setBook(bookById);
+        cartItem.setShoppingCart(new ShoppingCart(userId));
+        cartItemRepository.save(cartItem);
+        return getCart(authentication);
+    }
+
+    @Override
+    public ShoppingCartDto updateCartItem(
+            Long cartItemId,
+            UpdateCartItemsQuantityDto requestDto,
+            Authentication authentication
+    ) {
+        CartItem cartItemFromDb = cartItemRepository
+                .findById(cartItemId)
+                .orElseThrow(() -> new NoSuchElementException("Cart item not found"));
+        cartItemFromDb.setQuantity(requestDto.getQuantity());
+        cartItemRepository.save(cartItemFromDb);
+        return getCart(authentication);
+    }
+
+    @Override
+    public void deleteCartItem(Long cartItemId, Authentication authentication) {
+        Long userId = ((User) userDetailsService
+                .loadUserByUsername(authentication.getName())).getId();
+        CartItem existingCartItem = Optional.ofNullable(cartItemRepository
+                .findByIdAndByCartId(userId, cartItemId)).orElseThrow(
+                () -> new NoSuchElementException(
+                        "No such item found in your shopping cart"
+                )
+        );
```

Comment:
```suggestion
        CartItem existingCartItem = cartItemRepository
                .findByIdAndByCartId(userId, cartItemId).orElseThrow(
                () -> new EntityNotFoundException(
                        "No such item found in your shopping cart"
                )
        );
```


add parameters to message

Reply:
++, updated